### PR TITLE
[TRAV-192] increment version numbers when updating objects using doc API

### PR DESCRIPTION
**Pull request by clicktravel-andrew**
- Fixed a defect where version numbers were not being initialised and thereafter incremented when creating/updating document rows.
- Updating to add a version check to the update method to ensure correct versioning of data in Dynamo.
- Adding unit test
- Moved version constant to Abstract parent class for use by both children

@clicktravel-robin @clicktravel-basharat @clicktravel-tajinder @clicktravel-steffan - please review and merge if ok and we will do the framework bump.

### CODE REVIEW

File: cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/persistence/aws/dynamodb/DynamoDocumentStoreTemplate.java [hunk 1]
```diff
@@ -311,7 +316,7 @@ private PrimaryKey getPrimaryKey(final ItemId itemId, final ItemConfiguration it
         return value.toString();
     }
 
-    private <T extends Item> T stringToItem(final String item, final Class<T> valueType) {
+    <T extends Item> T stringToItem(final String item, final Class<T> valueType) {
```
Comment (clicktravel-robin) [inline cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/persistence/aws/dynamodb/DynamoDocumentStoreTemplate.java]: was this scope change intentional?

File: cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/persistence/aws/dynamodb/DynamoDocumentStoreTemplate.java [hunk 2]
```diff
@@ -166,22 +168,25 @@ public void initialize(final AmazonDynamoDB amazonDynamoDbClient) {
 
     @Override
     public <T extends Item> T update(final T item, final PersistenceExceptionHandler<?>... persistenceExceptionHandlers) {
-
         final ItemConfiguration itemConfiguration = getItemConfiguration(item.getClass());
-        final Item tmpItem = read(itemConfiguration.getItemId(item), item.getClass());
-        if (tmpItem == null) {
-            throw new NonExistentItemException(String.format("The document of type [%s] with id [%s] does not exist",
-                    item.getClass().getName(), itemConfiguration.getItemId(item)));
+        if (item.getVersion() == null) {
```
Comment (clicktravel-robin) [inline cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/persistence/aws/dynamodb/DynamoDocumentStoreTemplate.java]: I'm guessing this is what the other template does too

File: cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/persistence/aws/dynamodb/DynamoDocumentStoreTemplate.java (hunk 1 above)
Comment (clicktravel-andrew) [inline cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/persistence/aws/dynamodb/DynamoDocumentStoreTemplate.java]: yes, it allows the method to be used in the unit test for converting between items and JSON strings to set up the expectations of mock return values.

File: cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/persistence/aws/dynamodb/DynamoDocumentStoreTemplate.java (hunk 2 above)
Comment (clicktravel-andrew) [inline cheddar/cheddar-integration-aws/src/main/java/com/clicktravel/infrastructure/persistence/aws/dynamodb/DynamoDocumentStoreTemplate.java]: yeah, saves a read and follows the pre-established pattern.

File: cheddar/cheddar-integration-aws/src/test/java/com/clicktravel/infrastructure/persistence/aws/dynamodb/DynamoDocumentStoreTemplateTest.java [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+        final DynamoDocumentStoreTemplate dynamoDocumentStoreTemplate = new DynamoDocumentStoreTemplate(
+                mockDatabaseSchemaHolder);
+        dynamoDocumentStoreTemplate.initialize(mockAmazonDynamoDbClient);
+
+        final Item mockTableItem = mock(Item.class);
+        when(mockTableItem.toJSON()).thenReturn(dynamoDocumentStoreTemplate.itemToString(stubItem));
+
+        doThrow(RuntimeException.class).when(mockTable).putItem(any(PutItemSpec.class));
+
+        exception.expect(RuntimeException.class);
```
Comment (clicktravel-steffan) [inline cheddar/cheddar-integration-aws/src/test/java/com/clicktravel/infrastructure/persistence/aws/dynamodb/DynamoDocumentStoreTemplateTest.java]: So far in Cheddar we use a convention for verifying exceptions. Its is slightly more verbose, but we have favoured this approach for clarity:
```java
// Given
...

// When
MyException thrownException = null;
try {
   obj.method(); // method under test
} catch (MyException e) {
   thrownException = e;
}

// Then
assertNotNull(thrownException);
...
```

**Comment by clicktravel-steffan** [comment]
@clicktravel-andrew I've completed this review, just the one comment from me about checking thrown exceptions in unit tests.

**Comment by clicktravel-andrew** [comment]
Updated following review comments from @clicktravel-steffan

**Pull request merged**